CMakeLists.txt

```
@@ -105,6 +104,12 @@ if(DJINNI_WITH_JNI)
   set_target_properties(djinni_support_lib PROPERTIES
     POSITION_INDEPENDENT_CODE ON)
   target_include_directories(djinni_support_lib PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/djinni/jni/>")
+  # if this is a dynamic lib, we must have a main function in the lib
+  if ("${DJINNI_LIBRARY_TYPE}" STREQUAL "SHARED" )
+    list(APPEND SRC_JNI "djinni/jni/djinni_jni_main.cpp")
+  elseif(NOT "${DJINNI_LIBRARY_TYPE}" STREQUAL "STATIC" AND BUILD_SHARED_LIBS)
+    list(APPEND SRC_JNI "djinni/jni/djinni_jni_main.cpp")
+  endif()
```

Why is this two branches for the same case? Also maybe it's better to check if the variable is not defined instead of not being `STATIC` in the second case? While `DJINNI_LIBRARY_TYPE` should only be set to `SHARED` or `STATIC`, It may technically be possible to also set other library types? 🤔 
```suggestion
  # if this is a shared library, it should provide it's own `JNI_OnLoad` and `JNI_OnUnload` implementations.
  if ( DJINNI_LIBRARY_TYPE STREQUAL "SHARED" OR ( BUILD_SHARED_LIBS AND NOT DEFINED DJINNI_LIBRARY_TYPE ))
    list(APPEND SRC_JNI "djinni/jni/djinni_jni_main.cpp")
  endif()
```
Disclaimer: I haven't tested this!

---

Its less clever as I have it, but I find it more easy to read. 
The result is the same, so functionally there is no difference, and I do not mind looking not clever by doing it that way.

But if you strongly feel we should just have one branch and join the conditions init 1 long one, I also do not mind.
Shall I do it? 

You could set `DJINNI_LIBRARY_TYPE` to `MODULE` and it will not trigger an error, while everything else (beside STATIC/SHARED) should, but I do not want even to think about that :-)